mypy/checker.py

```diff
@@ -552,6 +602,16 @@ def check_func_def(self, defn: FuncItem, typ: CallableType, name: str) -> None:
                     self.fail(messages.RETURN_TYPE_CANNOT_BE_CONTRAVARIANT,
                          typ.ret_type)
 
+            # check that Generator functions have the appropriate return type
+            if defn.is_generator:
+                if not self.is_generator_return_type(typ.ret_type):
+                    self.fail(messages.INVALID_RETURN_TYPE_FOR_GENERATOR, typ)
+
+                # Python 2 generators aren't allowed to return values
+                if self.pyversion[0] == 2 and typ.ret_type.type.fullname() == 'typing.Generator':
```

Comment:
I wasn't sure how to write a unit test for this, but it should probably have one.